In [2]:
import pandas as pd
data = pd.read_csv("flipkard.csv")
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_id          80000 non-null  str    
 1   product_name        80000 non-null  str    
 2   category            80000 non-null  str    
 3   brand               80000 non-null  str    
 4   seller              80000 non-null  str    
 5   seller_city         80000 non-null  str    
 6   price               80000 non-null  float64
 7   discount_percent    80000 non-null  int64  
 8   final_price         80000 non-null  float64
 9   rating              80000 non-null  float64
 10  review_count        80000 non-null  int64  
 11  stock_available     80000 non-null  int64  
 12  units_sold          80000 non-null  int64  
 13  listing_date        80000 non-null  str    
 14  delivery_days       80000 non-null  int64  
 15  weight_g            80000 non-null  float64
 16  warranty_months

In [8]:
data.isnull().sum()
data['size'] =  data['size'].fillna(0)


In [9]:
data.isnull().sum()

product_id            0
product_name          0
category              0
brand                 0
seller                0
seller_city           0
price                 0
discount_percent      0
final_price           0
rating                0
review_count          0
stock_available       0
units_sold            0
listing_date          0
delivery_days         0
weight_g              0
warranty_months       0
color                 0
size                  0
return_policy_days    0
is_returnable         0
payment_modes         0
shipping_weight_g     0
product_score         0
seller_rating         0
dtype: int64

In [19]:
discounted_price = data['price']-(data['price']*(data['discount_percent']/100))
discounted_price

0        30215.2390
1        27624.4110
2        30728.7000
3        23217.9430
4        14181.3400
            ...    
79995    40782.1985
79996    18142.6950
79997    20254.1600
79998     4448.7060
79999    38831.0650
Length: 80000, dtype: float64

In [25]:
data['Total_revenue'] = discounted_price * data['units_sold']
data['Total_revenue'].sum()

np.float64(4761907957160.211)

In [26]:
data.groupby('category')['Total_revenue'].sum().sort_values(ascending=False)


category
Toys              6.064486e+11
Beauty            6.058160e+11
Fashion           5.995559e+11
Electronics       5.994879e+11
Appliances        5.940603e+11
Sports            5.890503e+11
Mobiles           5.843309e+11
Home & Kitchen    5.831580e+11
Name: Total_revenue, dtype: float64

key insights:
- toys generate the highest total revenue, followed closely by beauty and fashion.
- the revenue difference between categories is relatively small.
- this indicates that revenue is evenly distributed across categories , with no single category overwhelmingly dominating sales.
- the bsuiness appears to have a balanced revenue structure across product segments

In [27]:
data.columns

Index(['product_id', 'product_name', 'category', 'brand', 'seller',
       'seller_city', 'price', 'discount_percent', 'final_price', 'rating',
       'review_count', 'stock_available', 'units_sold', 'listing_date',
       'delivery_days', 'weight_g', 'warranty_months', 'color', 'size',
       'return_policy_days', 'is_returnable', 'payment_modes',
       'shipping_weight_g', 'product_score', 'seller_rating', 'Total_revenue'],
      dtype='str')

In [3]:
data['product_revenue'] = data['final_price'] * data['units_sold']
data.groupby('product_name')['product_revenue'].sum().sort_values(ascending=False).head(10)

product_name
HP Series 622            7.470177e+08
Apple Prime 418          6.894918e+08
Philips Ultra 569        6.417605e+08
Puma Edition 70          6.372428e+08
HP Series 569            6.188757e+08
Prestige Series 377      6.167213e+08
Whirlpool Edition 621    6.109708e+08
Apple Series 993         6.006041e+08
LG Series 980            5.952644e+08
LG Ultra 138             5.946968e+08
Name: product_revenue, dtype: float64

key insights:

- The highest revenue generating product is HP Series 622 .
- Other high-performing products include Apple , Prime 418 and Phillips ultra 569
- Revenue concentration appears among premium branded products .
- Electronics and branded appliances dominate the top revenue list.

Business insight:
High revenue products are likely driven by :
- Strong brand reputation
- Higher price points
- Higher units sold

These products should be prioritized for 
- Marketing campaigns
- Inventory Optimization
-Bundle offers

In [30]:
seller_revenue = data.groupby('seller')['product_revenue'].sum().sort_values(ascending= False).head(10)
seller_revenue

seller
UrbanRetails    6.043700e+11
QuickShop       6.014686e+11
BestBuy         5.964748e+11
SuperMart       5.953379e+11
ValueKart       5.948731e+11
MegaStore       5.937561e+11
SmartDeals      5.881427e+11
RetailHub       5.874846e+11
Name: product_revenue, dtype: float64

Seller Revenue Analysis

- UrbanRetails generate the highest total revenue.
- Followed by Quickshop and BestBuy
- Revenue distribution across top sellers is relatively close , indicating competitive performance.

Business Insight:
High - performing sellers may have :
- Better product mix
- Higher demand categories
- Better pricing strategy
- Strong ratings or faster delivery

In [7]:
brand_revenue = data.groupby('brand')['product_revenue'].sum().sort_values(ascending=False)
brand_revenue.head(10)


brand
Nike        3.237334e+11
Adidas      3.213176e+11
Puma        3.210516e+11
Apple       3.209929e+11
Boat        3.206118e+11
Prestige    3.196137e+11
HP          3.183560e+11
Dell        3.177141e+11
Reebok      3.171473e+11
LG          3.167956e+11
Name: product_revenue, dtype: float64

In [8]:
brand_by_unitSold = data.groupby('brand')['units_sold'].sum().sort_values(ascending=False)
brand_by_unitSold.head(10)

brand
Nike        13569204
Reebok      13507145
Prestige    13497983
Adidas      13459726
Puma        13443146
Apple       13425297
Dell        13415026
LG          13403447
HP          13356680
Redmi       13320362
Name: units_sold, dtype: int64

In [9]:
revenue_set = set(brand_revenue.index)
unit_set = set(brand_by_unitSold.index)
common_brands = revenue_set & unit_set
only_revenue = revenue_set - unit_set
print("Brands with both revenue and units sold data:", common_brands)
print("Brands with only revenue data:", only_revenue)

Brands with both revenue and units sold data: {'Whirlpool', 'Reebok', 'Adidas', 'Philips', 'Boat', 'Dell', 'Puma', 'Nike', 'Redmi', 'HP', 'Samsung', 'Sony', 'Prestige', 'LG', 'Apple'}
Brands with only revenue data: set()


Brand Performance Analysis

Brands appearing in both top revenue and top units sold lists represent true market leaders.  
These brands generate high income while also maintaining strong sales volume.

No brands were found that rank high in revenue but not in units sold, indicating that revenue is primarily driven by volume rather than premium pricing.

This suggests the marketplace favors mass-market products over niche luxury items.


In [10]:
category_unit_sold = data.groupby('category')['units_sold'].sum().sort_values(ascending=False)
category_unit_sold.head(10)

category
Toys              25565969
Electronics       25372487
Appliances        25184186
Fashion           25182630
Beauty            25129312
Sports            24792347
Mobiles           24710378
Home & Kitchen    24663953
Name: units_sold, dtype: int64

In [12]:
data['price_range'] = pd.cut(data['final_price'],bins=5,labels=['Low','Medium','High','Very High','Premium'])
price_demand = data.groupby('price_range')['units_sold'].sum().sort_values(ascending=False)
price_demand

price_range
Medium       53442009
Low          53441811
High         48578015
Very High    31731893
Premium      13407534
Name: units_sold, dtype: int64

customers prefer mid-priced products over premium ones.
sales volume decreases as price increases, indicating price-sensitive demand.
low and medium price ranges contribute the highest number of units sold, while premium products sell the least.

In [14]:
data['reviews'] = pd.cut(data['rating'], bins=5, labels=['Very Low', 'Low', 'Medium', 'High', 'Very High'])
review_demand = data.groupby('reviews')['units_sold'].sum().sort_values(ascending=False)
review_demand

reviews
Very Low     42736569
Low          40269277
Medium       39972899
High         39957305
Very High    37665212
Name: units_sold, dtype: int64

In [16]:
data['rating_revenue'] = pd.cut(data['rating'], bins=5, labels=['Very Low', 'Low', 'Medium', 'High', 'Very High'])
rating_revenue = data.groupby('rating_revenue')['product_revenue'].sum().sort_values(ascending=False)
rating_revenue

rating_revenue
Very Low     1.017723e+12
Low          9.569223e+11
Medium       9.546686e+11
High         9.445581e+11
Very High    8.880360e+11
Name: product_revenue, dtype: float64

In [18]:
sales_by_deliveryspeed = data.groupby('delivery_days')['units_sold'].sum().sort_values(ascending=False)
sales_by_deliveryspeed

delivery_days
7     18621115
4     18587030
1     18560463
11    18468055
6     18249885
10    18163012
5     18161685
9     18142833
8     18092809
3     17981905
2     17572470
Name: units_sold, dtype: int64

Impact of Delivery Time on Sales

- products with approximately 7 days delivery generate the highest total sales volume.
- faster delivery does not necessarily lead to higher demand.
- this suggests that standard delivery times are acceptable for most customers.
- sales performance may depend more on product value than delivery speed.

total sales are influnced by the number of products offered with each delivery duration,not solely by customer preference.

In [19]:
data.columns

Index(['product_id', 'product_name', 'category', 'brand', 'seller',
       'seller_city', 'price', 'discount_percent', 'final_price', 'rating',
       'review_count', 'stock_available', 'units_sold', 'listing_date',
       'delivery_days', 'weight_g', 'warranty_months', 'color', 'size',
       'return_policy_days', 'is_returnable', 'payment_modes',
       'shipping_weight_g', 'product_score', 'seller_rating',
       'product_revenue', 'brand_revenue', 'price_range', 'reviews',
       'rating_revenue'],
      dtype='str')

In [20]:
data.groupby('return_policy_days')['units_sold'].sum().sort_values(ascending = False)

return_policy_days
10    40536489
15    40373574
7     40328012
0     39957883
30    39405304
Name: units_sold, dtype: int64

impact of return policy on sales 

- Products offering a moderate return period (around 10–15 days) achieve the highest sales.
- Extremely short return windows may discourage purchases due to perceived risk.
- Very long return periods do not significantly increase demand and may increase operational costs.
- A balanced return policy appears optimal for maximizing sales while controlling returns.

return policies influence customer trust , but beyond a certain point , extending the return period yields diminishing returns on sales

In [21]:
data.columns

Index(['product_id', 'product_name', 'category', 'brand', 'seller',
       'seller_city', 'price', 'discount_percent', 'final_price', 'rating',
       'review_count', 'stock_available', 'units_sold', 'listing_date',
       'delivery_days', 'weight_g', 'warranty_months', 'color', 'size',
       'return_policy_days', 'is_returnable', 'payment_modes',
       'shipping_weight_g', 'product_score', 'seller_rating',
       'product_revenue', 'brand_revenue', 'price_range', 'reviews',
       'rating_revenue'],
      dtype='str')

In [22]:
data.groupby('seller_rating')['units_sold'].sum().sort_values(ascending=False)

seller_rating
4.67    1155159
3.22    1134631
3.81    1125016
4.87    1124984
4.07    1123780
         ...   
4.94     887064
4.54     870971
3.30     838120
3.00     502495
5.00     429125
Name: units_sold, Length: 201, dtype: int64

Impact of Seller Rating on Sales

- Products sold by highly rated sellers achieve significantly higher sales volumes.
- Customer trust strongly influences purchasing decisions in online marketplaces.
- Lower-rated sellers experience reduced demand, likely due to perceived risk.
- Maintaining high seller ratings is critical for sustained sales performance.


In [23]:
data.groupby('stock_available')['units_sold'].sum().sort_values(ascending=False)

stock_available
125    286979
322    281474
194    278082
68     277285
239    274258
        ...  
618    143052
238    133672
898    132716
10     130839
187    125294
Name: units_sold, Length: 1000, dtype: int64

Stock Availability vs Sales

- Products with higher available stock tend to record greater sales volumes.
- High inventory levels often indicate strong anticipated demand.
- Low stock availability may constrain potential sales due to stockouts.
- Effective inventory management is critical to maximize revenue and avoid lost sales

In [25]:
data['score_range'] = pd.cut(data['product_score'],bins=5,labels=['Very Low','Low','Medium','High','Very High'])
data.groupby('score_range')['units_sold'].sum().sort_values(ascending=False)


score_range
Medium       40708255
Very High    40500479
Low          40148575
High         39747810
Very Low     39496143
Name: units_sold, dtype: int64

Impact of Product Rating on Sales

Products with medium ratings generate the highest sales volume, 
indicating that customers prioritize affordability and availability 
over perfect ratings.

Very low-rated products perform the worst, showing that extremely 
poor quality strongly reduces demand.

Highly rated products do not necessarily achieve the highest sales, 
likely because they belong to premium price segments.

In [26]:
score_demand = data.groupby('product_score')['units_sold'].sum().sort_values(ascending=False)
score_demand

product_score
86.61    59412
39.15    55019
29.93    54669
8.19     54651
27.33    52613
         ...  
10.47     1020
19.74      982
35.69      638
41.70      571
38.84      246
Name: units_sold, Length: 9895, dtype: int64

In [27]:
correlation = data.corr(numeric_only=True)
correlation

price  discount_percent  final_price    rating  \
price               1.000000          0.001764     0.922674 -0.003839   
discount_percent    0.001764          1.000000    -0.333058  0.006917   
final_price         0.922674         -0.333058     1.000000 -0.005528   
rating             -0.003839          0.006917    -0.005528  1.000000   
review_count        0.003205         -0.003364     0.005294 -0.002284   
stock_available     0.003740         -0.002199     0.004056  0.003924   
units_sold          0.004065         -0.000088     0.004778 -0.001172   
delivery_days      -0.007037          0.006568    -0.008907 -0.003450   
weight_g           -0.000733         -0.000648    -0.000762  0.004408   
warranty_months    -0.000316          0.002192    -0.000133 -0.006851   
return_policy_days  0.009869          0.008013     0.007286 -0.001171   
is_returnable       0.008128          0.006500     0.006173  0.002173   
shipping_weight_g  -0.000575         -0.001110    -0.000453  0.004126   
product_score      -0.000966          0.004253    -0.002668 -0.004429   
seller_rating       0.003173         -0.001854     0.004096 -0.000825   
product_revenue     0.624792         -0.227555     0.678805 -0.004910   
brand_revenue            NaN               NaN          NaN       NaN   

                    review_count  stock_available  units_sold  delivery_days  \
price                   0.003205         0.003740    0.004065      -0.007037   
discount_percent       -0.003364        -0.002199   -0.000088       0.006568   
final_price             0.005294         0.004056    0.004778      -0.008907   
rating                 -0.002284         0.003924   -0.001172      -0.003450   
review_count            1.000000        -0.002444    0.002285       0.000355   
stock_available        -0.002444         1.000000   -0.006276      -0.004676   
units_sold              0.002285        -0.006276    1.000000       0.001608   
delivery_days           0.000355        -0.004676    0.001608       1.000000   
weight_g                0.003814        -0.000410   -0.001319       0.006493   
warranty_months         0.002071         0.001440    0.003501      -0.000319   
return_policy_days     -0.003098        -0.005026   -0.003865       0.001433   
is_returnable           0.003081        -0.009533    0.001675      -0.004216   
shipping_weight_g       0.004445        -0.000598   -0.002128       0.006651   
product_score           0.002515         0.002638   -0.001704       0.001048   
seller_rating           0.001907         0.000119   -0.003513       0.000868   
product_revenue         0.004266        -0.000065    0.627344      -0.005307   
brand_revenue                NaN              NaN         NaN            NaN   

                    weight_g  warranty_months  return_policy_days  \
price              -0.000733        -0.000316            0.009869   
discount_percent   -0.000648         0.002192            0.008013   
final_price        -0.000762        -0.000133            0.007286   
rating              0.004408        -0.006851           -0.001171   
review_count        0.003814         0.002071           -0.003098   
stock_available    -0.000410         0.001440           -0.005026   
units_sold         -0.001319         0.003501           -0.003865   
delivery_days       0.006493        -0.000319            0.001433   
weight_g            1.000000        -0.005982            0.000167   
warranty_months    -0.005982         1.000000           -0.004262   
return_policy_days  0.000167        -0.004262            1.000000   
is_returnable      -0.000567        -0.003098            0.616076   
shipping_weight_g   0.994880        -0.005471            0.000490   
product_score      -0.003932         0.000660           -0.002488   
seller_rating      -0.002936        -0.005818           -0.000008   
product_revenue    -0.001724         0.001217            0.000781   
brand_revenue            NaN              NaN                 NaN   

                    is_r

Correlation analysis shows that price and final_price are strongly 
positively correlated, as expected, while discount percentage has a 
moderate negative relationship with final_price.

Most other variables, including rating, review count, delivery time, 
and stock availability, show very weak correlations with units sold. 
This suggests that sales are influenced by multiple factors rather 
than a single variable

In [28]:
units_corr = data.corr(numeric_only=True)['units_sold'].sort_values(ascending=False)
units_corr

units_sold            1.000000
product_revenue       0.627344
final_price           0.004778
price                 0.004065
warranty_months       0.003501
review_count          0.002285
is_returnable         0.001675
delivery_days         0.001608
discount_percent     -0.000088
rating               -0.001172
weight_g             -0.001319
product_score        -0.001704
shipping_weight_g    -0.002128
seller_rating        -0.003513
return_policy_days   -0.003865
stock_available      -0.006276
brand_revenue              NaN
Name: units_sold, dtype: float64

Correlation analysis with units_sold shows that product_revenue has a 
strong positive relationship with sales, which is expected since revenue 
depends directly on units sold.

All other variables exhibit very weak correlations, indicating that 
no single factor such as price, rating, delivery time, or discount 
alone strongly determines sales performance. This suggests that sales 
are influenced by multiple combined factors rather than a single driver.

The analysis reveals that mid-priced products, especially in Toys and 
Electronics categories, generate the highest sales volume. Price-sensitive 
demand is evident as sales decrease in premium price ranges.

Customer ratings alone do not strongly influence sales, suggesting that 
other factors such as price positioning, product type, and availability 
play a more significant role.

Faster delivery and flexible return policies show slight positive impact 
on sales, indicating the importance of customer convenience.

Businesses should prioritize inventory and marketing efforts toward 
high-performing categories and optimize pricing strategies to maximize 
revenue and demand.